File: 146.LRU Cache/lru_cache.cpp
class LRUCache {
public:
    class Node {
        public: 
        int key, value;
        Node* next, *prev;
        Node(int key, int value) {
            this->key = key; 
            this->value = value; 
            next = NULL; 
            prev = NULL; 
        }
    }; 

    int capacity; 
    unordered_map<int, Node*> cache; 

    Node* left = new Node(-1, -1); 
    Node* right = new Node(-1, -1);

    LRUCache(int capacity) {
        this->capacity = capacity;
        left->next = right;
        right->prev = left; 
        
    }

    void deleteNode(Node* node) {
        Node* prev = node->prev; 
        Node* next = node->next;

        prev->next = next; 
        next->prev = prev; 
    }

    void insertRight(Node* node) {
        Node* temp = right->prev;
        temp->next = node; 
        node->prev = temp; 

        node->next = right; 
        right->prev = node; 
    }
    
    int get(int key) {
        if (cache.find(key) == cache.end()) {
            return -1;
        }
        Node* currNode = cache[key];
        deleteNode(currNode); 
        insertRight(currNode); 
        
        return currNode->value; 
    }
    
    void put(int key, int value) {
        //check if in map 
        if (cache.find(key) != cache.end()) {
            Node* currNode = cache[key]; 
            deleteNode(currNode);
            Node* updatedNode = new Node(key, value); 
            insertRight(updatedNode);
            cache[key] = updatedNode; 
        }
        else {
            if(cache.size() == capacity) {
                //delete left end node 
                //insert right 
                Node* leftNode = left->next; 

                cache.erase(leftNode->key);
                deleteNode(leftNode); 
                Node* newNode = new Node(key, value); 
                cache[key] = newNode; 
                insertRight(newNode); 
            }
            else {
                //no need to delete 
                Node* newNode = new Node(key, value); 
                cache[key] = newNode; 
                insertRight(newNode); 
            }
        }

        //if not in map, check if full 
        //if full, pop
        //insert 
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146.LRU Cache/lru_cache.py
class Node:
    def __init__(self, key, val):
        self.key, self.val = key, val 
        self.prev = self.next = None 

class LRUCache(object):

    def __init__(self, capacity):
        self.cap = capacity 
        self.cache = {}
        self.left, self.right = Node(0, 0), Node(0, 0)
        self.left.next, self.right.prev = self.right, self.left 
        """
        :type capacity: int
        """
        
    def remove(self, node):
        prev, nxt = node.prev, node.next 
        nxt.prev, prev.next = prev, nxt
    
    def insert(self, node): 
        right = self.right 
        prev = right.prev 
        prev.next, right.prev = node, node
        node.next, node.prev = right, prev


    def get(self, key):
        if key in self.cache:
            self.remove(self.cache[key])
            self.insert(self.cache[key])
            return self.cache[key].val
        return -1
        """
        :type key: int
        :rtype: int
        """
        

    def put(self, key, value):
        if key in self.cache: 
            self.remove(self.cache[key])
        self.cache[key] = Node(key, value)
        self.insert(self.cache[key])

        if len(self.cache) > self.cap:
            #remove and delete LRU from hashmap
            lru = self.left.next 
            self.remove(lru)
            del self.cache[lru.key]


        """
        :type key: int
        :type value: int
        :rtype: None
        """
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)